allay/plugin.py

```diff
@@ -1,23 +1,52 @@
+import json
+from typing import Union, cast
+
 from beet import Context
+from beet.contrib.messages import Message
 from beet.core.container import Drop
 from beet.core.file import TextFile
-from beet.contrib.messages import Message
+from beet.core.utils import JsonDict
 from beet.library.base import NamespaceFile
 from beet.library.data_pack import DataPack
 
 from .parser import Parser
 
+parser = Parser()
+
 
 class AllayMessage(TextFile, NamespaceFile):
     scope = ("messages",)
     extension = ".allay"
 
     def bind(self, pack: DataPack, path: str):
-        json = Parser().parse(self.text)
+        json = parser.parse(self.text)
         pack[path] = Message(json)
 
         raise Drop()
 
 
+def register_pattern(name: str, raw: Union[JsonDict, str]):
+    if isinstance(raw, str):
+        raw = json.loads(parser.parse(f'@{name} = {raw}\n#ALLAYDEFS\n'))
+    else:
+        parser.patterns[f"@{name}"] = raw
+
+
+def register_template(name: str, raw: Union[JsonDict, str]):
+    if isinstance(raw, str):
+        raw = json.loads(parser.parse(raw))
```

Comment:
Templates also are required to be inside the definitions block, so I think you need to add that as well. Shouldn't `register_pattern` and `register_template` be nearly identical?